integration/util.py

```diff
@@ -539,21 +569,32 @@ def __init__(self, magic_text):
         self.exited = Deferred()
         self._magic_text = magic_text
         self._output = StringIO()
-        self._eliot_stream = EliotLogStream()
+        self._eliot_stream = EliotLogStream(fallback=self.eliot_garbage_received)
+        self._eliot_stderr = EliotLogStream(fallback=self.err_received)
         self._action = current_action()
         assert self._action is not None
 
     def processEnded(self, reason):
         with self._action:
             Message.log(message_type=u"process-ended")
+        if not self.magic_seen.called:
```

Comment:
I generally think that application-level inspection of `Deferred.called` indicates something has gone fairly wrong.

Perhaps this particular code looks like it's okay ... it is at least the creator/caller of the Deferred that is doing the checking.  That basically means it's just a shortcut to avoid having to keep track of whether this same unit-of-responsibility has fired the Deferred or not yet.

I still think it would be better to track this information some other way.  Perhaps by dropping `magic_seen` (setting it to `None`) immediately before firing it and then checking it against `None`?

I think it's worth doing this because even though this instance of looking at `.called` is probably okay, I'd rather have a codebase where nothing ever looks at `.called` and not have to stop and think about whether certain uses are okay or not.